jupyterhub/tests/conftest.py

```diff
@@ -322,26 +325,26 @@ def cleanup():
             with raises(TimeoutExpired):
                 service.proc.wait(1)
         if url:
-            io_loop.run_sync(partial(service.server.wait_up, http=True))
+            await service.server.wait_up(http=True)
     return service
 
 
 @fixture
-def mockservice(request, app):
+async def mockservice(request, app):
     """Mock a service with no external service url"""
-    yield _mockservice(request, app, url=False)
+    yield await _mockservice(request, app, url=False)
```

Comment:
I have not encountered a scenario where `yield` and `await` are used in this way. Can you please help me understand the usage?

Reply:
Absolutely! Fixtures are used to do one or both of:

1. set something up before the test
2. clean up after the test

If you only need to do setup, the fixture function can `return` what should be the value of the fixture (if any) in tests that use it.

If you need to _cleanup_, then you can `yield` what should be passed to the functions (can be nothing, if it's purely setup/teardown and has no object to be passed to tests). You can read more in the [pytest docs on fixture teardown/cleanup](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#yield-fixtures-recommended).

So a fixture with setup _and_ teardown generally looks like:

```python
@pytest.fixture
def my_fixture():
    setup_before()
    yield value_for_tests_if_any
    cleanup_after()
```

The `await` doesn't change anything here. When you `yield <expression>`, what's yielded is the _result_ of the expression, which means the `awaited` result if there's an `await` in there. That means

```python
yield await async_thing
```

is the same as

```python
thing = await async_thing
yield thing
```

if that makes things clearer.


If you run [this sample test](https://gist.github.com/d6fa0aaaa35ec8bad55db5136499c728) with:

```
pytest -vs example.py
```

you'll get:

```
example.py::test_sample in setup
in async function
myfixture='sync result'
PASSEDin teardown
```

Reply:
I definitely understand this better. Thank you for the explanation.
When I run the code I get this though:
```python
example.py::test_sample <async_generator object myfixture at 0x7f22b8408af0>
PASSED
```

Reply:
Forgot about my global config. It should be:

```
pytest -vs --asyncio-mode auto example.py

Reply:
Yeah! This works now. Thank you!